File: backend/app/services/latex_generator.py
"""
LaTeX Systematic Literature Review Generator

Generates LaTeX documents for systematic literature reviews using Gemini AI
to write content based on collected papers.
"""

import logging
import os
from typing import List, Dict, Optional
from datetime import datetime
import google.generativeai as genai

from app.core.config import settings
from app.services.bibtex_generator import BibTeXGenerator

logger = logging.getLogger(__name__)


class LaTeXReviewGenerator:
    """Generates systematic literature review documents in LaTeX."""

    def __init__(self):
        """Initialize the generator with Gemini API."""
        self.bibtex_gen = BibTeXGenerator()

        # Configure Gemini
        if settings.GEMINI_API_KEY:
            genai.configure(api_key=settings.GEMINI_API_KEY)
            self.model = genai.GenerativeModel('gemini-pro')
        else:
            logger.warning("GEMINI_API_KEY not configured, AI generation disabled")
            self.model = None

    def _generate_latex_header(self, title: str, author: str = "Anonymous") -> str:
        """Generate LaTeX document header."""
        return f"""\\documentclass[12pt,a4paper]{{article}}

% Packages
\\usepackage[utf8]{{inputenc}}
\\usepackage[T1]{{fontenc}}
\\usepackage{{geometry}}
\\usepackage{{cite}}
\\usepackage{{url}}
\\usepackage{{hyperref}}
\\usepackage{{graphicx}}
\\usepackage{{booktabs}}
\\usepackage{{longtable}}

% Page geometry
\\geometry{{
    a4paper,
    left=2.5cm,
    right=2.5cm,
    top=2.5cm,
    bottom=2.5cm
}}

% Hyperref setup
\\hypersetup{{
    colorlinks=true,
    linkcolor=blue,
    citecolor=blue,
    urlcolor=blue
}}

% Title information
\\title{{{title}}}
\\author{{{author}}}
\\date{{\\today}}

\\begin{{document}}

\\maketitle

\\begin{{abstract}}
This systematic literature review was automatically generated using LitRevTool.
The review analyzes papers collected from Google Scholar based on specific search criteria.
The PRISMA methodology was followed to ensure systematic and transparent reporting.
\\end{{abstract}}

\\tableofcontents
\\newpage
"""

    def _generate_introduction_section(self, search_criteria: Dict) -> str:
        """Generate introduction section with search methodology."""
        keywords_inc = ", ".join(search_criteria.get('keywords_include', []))
        keywords_exc = ", ".join(search_criteria.get('keywords_exclude', []))
        start_year = search_criteria.get('start_year', 'N/A')
        end_year = search_criteria.get('end_year', 'N/A')

        intro = f"""\\section{{Introduction}}

This systematic literature review examines the research literature based on the following search strategy:

\\subsection{{Search Criteria}}

\\begin{{itemize}}
    \\item \\textbf{{Inclusion Keywords}}: {keywords_inc}
"""

        if keywords_exc:
            intro += f"    \\item \\textbf{{Exclusion Keywords}}: {keywords_exc}\n"

        intro += f"""    \\item \\textbf{{Year Range}}: {start_year} to {end_year}
    \\item \\textbf{{Database}}: Google Scholar
\\end{{itemize}}

\\subsection{{Methodology}}

This review follows the PRISMA (Preferred Reporting Items for Systematic Reviews and Meta-Analyses) guidelines.
The search was conducted using LitRevTool, which overcomes Google Scholar's limitations by splitting searches by year.

"""
        return intro

    def _generate_prisma_section(self, prisma_metrics: Dict) -> str:
        """Generate PRISMA flow section with metrics table."""
        if not prisma_metrics:
            return ""

        identification = prisma_metrics.get('identification', {})
        screening = prisma_metrics.get('screening', {})
        eligibility = prisma_metrics.get('eligibility', {})
        included = prisma_metrics.get('included', {})

        return f"""\\section{{PRISMA Flow}}

The following table summarizes the systematic review process according to PRISMA guidelines:

\\begin{{table}}[h]
\\centering
\\begin{{tabular}}{{ll}}
\\toprule
\\textbf{{Stage}} & \\textbf{{Count}} \\\\
\\midrule
Records identified & {identification.get('records_identified', 0)} \\\\
Duplicates removed & {screening.get('records_excluded_duplicates', 0)} \\\\
Records screened & {screening.get('records_screened', 0)} \\\\
Papers assessed for eligibility & {eligibility.get('full_text_assessed', 0)} \\\\
Papers excluded (semantic criteria) & {eligibility.get('full_text_excluded_semantic', 0)} \\\\
\\midrule
\\textbf{{Studies included}} & \\textbf{{{included.get('studies_included', 0)}}} \\\\
\\bottomrule
\\end{{tabular}}
\\caption{{PRISMA flow summary}}
\\label{{tab:prisma}}
\\end{{table}}

"""

    def _generate_literature_review_with_ai(self, papers: List[Dict], search_criteria: Dict) -> str:
        """Use Gemini AI to generate literature review content."""
        if not self.model or not papers:
            return self._generate_basic_literature_review(papers)

        try:
            # Prepare paper summaries for AI
            paper_summaries = []
            for i, paper in enumerate(papers[:50]):  # Limit to first 50 for API constraints
                title = paper.get('title', 'Unknown')
                authors = paper.get('authors', 'Unknown')
                year = paper.get('year', 'N/A')
                abstract = paper.get('abstract', '')[:200]  # Truncate long abstracts

                paper_summaries.append(f"{i+1}. {title} ({authors}, {year})")
                if abstract:
                    paper_summaries.append(f"   Abstract: {abstract}...")

            papers_text = "\n".join(paper_summaries[:30])  # Further limit for prompt

            keywords = ", ".join(search_criteria.get('keywords_include', []))

            prompt = f"""Based on the following research papers about "{keywords}", write a comprehensive literature review section for an academic paper.

The review should:
1. Identify main themes and trends
2. Group related papers together
3. Highlight key findings and methodologies
4. Note any gaps or areas for future research
5. Be written in formal academic style
6. Be 400-600 words

Papers:
{papers_text}

Write the literature review section:"""

            response = self.model.generate_content(prompt)

            if response and response.text:
                return f"""\\section{{Literature Review}}

{response.text}

\\subsection{{Paper Distribution}}

A total of {len(papers)} papers were included in this review. The papers span various topics related to {keywords}.

"""
            else:
                return self._generate_basic_literature_review(papers)

        except Exception as e:
            logger.error(f"Error generating AI literature review: {e}")
            return self._generate_basic_literature_review(papers)

    def _generate_basic_literature_review(self, papers: List[Dict]) -> str:
        """Generate basic literature review without AI."""
        if not papers:
            return "\\section{Literature Review}\n\nNo papers were collected.\n\n"

        # Group papers by year
        papers_by_year = {}
        for paper in papers:
            year = paper.get('year', 'Unknown')
            if year not in papers_by_year:
                papers_by_year[year] = []
            papers_by_year[year].append(paper)

        review = "\\section{Literature Review}\n\n"
        review += f"A total of {len(papers)} papers were included in this systematic review.\n\n"

        review += "\\subsection{Temporal Distribution}\n\n"
        review += "The following table shows the distribution of papers by year:\n\n"
        review += "\\begin{table}[h]\n\\centering\n"
        review += "\\begin{tabular}{cc}\n\\toprule\n"
        review += "\\textbf{Year} & \\textbf{Count} \\\\\n\\midrule\n"

        for year in sorted(papers_by_year.keys(), reverse=True):
            review += f"{year} & {len(papers_by_year[year])} \\\\\n"

        review += "\\bottomrule\n\\end{tabular}\n"
        review += "\\caption{Papers by publication year}\n"
        review += "\\end{table}\n\n"

        return review

    def _generate_references_section(self) -> str:
        """Generate references section."""
        return """\\section{References}

\\bibliographystyle{plain}
\\bibliography{references}

"""

    def _generate_conclusion_section(self, paper_count: int) -> str:
        """Generate conclusion section."""
        return f"""\\section{{Conclusion}}

This systematic literature review identified and analyzed {paper_count} papers relevant to the specified search criteria.
The review followed PRISMA guidelines to ensure systematic and transparent reporting.

This document was automatically generated by LitRevTool, providing a foundation for further manual analysis and synthesis.

"""

    def generate_latex_document(
        self,
        papers: List[Dict],
        search_criteria: Dict,
        prisma_metrics: Dict = None,
        output_path: str = None,
        title: str = "Systematic Literature Review",
        author: str = "Generated by LitRevTool"
    ) -> str:
        """
        Generate complete LaTeX document.

        Args:
            papers: List of paper dictionaries
            search_criteria: Search parameters used
            prisma_metrics: PRISMA metrics dictionary
            output_path: Optional path to save the file
            title: Document title
            author: Document author

        Returns:
            Complete LaTeX content as string
        """
        latex_content = []

        # Header
        latex_content.append(self._generate_latex_header(title, author))

        # Introduction with search methodology
        latex_content.append(self._generate_introduction_section(search_criteria))

        # PRISMA section
        if prisma_metrics:
            latex_content.append(self._generate_prisma_section(prisma_metrics))

        # Literature review (AI-generated if possible)
        latex_content.append(self._generate_literature_review_with_ai(papers, search_criteria))

        # Conclusion
        latex_content.append(self._generate_conclusion_section(len(papers)))

        # References
        latex_content.append(self._generate_references_section())

        # End document
        latex_content.append("\\end{document}")

        full_content = "\n".join(latex_content)

        # Save to file if path provided
        if output_path:
            try:
                with open(output_path, 'w', encoding='utf-8') as f:
                    f.write(full_content)
                logger.info(f"LaTeX document saved to {output_path}")
            except Exception as e:
                logger.error(f"Error saving LaTeX document: {e}")

        return full_content


def generate_systematic_review(
    papers: List[Dict],
    search_criteria: Dict,
    prisma_metrics: Dict = None,
    latex_output_path: str = None,
    bibtex_output_path: str = None,
    title: str = "Systematic Literature Review"
) -> tuple:
    """
    Generate both LaTeX document and BibTeX file for systematic review.

    Args:
        papers: List of paper dictionaries
        search_criteria: Search parameters
        prisma_metrics: PRISMA metrics
        latex_output_path: Path to save LaTeX file
        bibtex_output_path: Path to save BibTeX file
        title: Document title

    Returns:
        Tuple of (latex_content, bibtex_content)
    """
    # Generate LaTeX
    latex_gen = LaTeXReviewGenerator()
    latex_content = latex_gen.generate_latex_document(
        papers=papers,
        search_criteria=search_criteria,
        prisma_metrics=prisma_metrics,
        output_path=latex_output_path,
        title=title
    )

    # Generate BibTeX
    bibtex_gen = BibTeXGenerator()
    bibtex_content = bibtex_gen.generate_bibtex_file(
        papers=papers,
        output_path=bibtex_output_path
    )

    return latex_content, bibtex_content

File: backend-node/src/services/latexGenerator.ts
/**
 * LaTeX Systematic Literature Review Generator
 * Generates LaTeX documents for systematic literature reviews using Gemini AI
 */
import fs from 'fs';
import { GoogleGenerativeAI } from '@google/generative-ai';
import { config } from '../core/config';
import logger from '../core/logger';
import { BibTeXGenerator } from './bibtexGenerator';

interface Paper {
  title: string;
  authors?: string;
  year?: number;
  abstract?: string;
  url?: string;
  source?: string;
  publisher?: string;
  citations?: number;
  [key: string]: any;
}

interface SearchCriteria {
  keywords_include?: string[];
  keywords_exclude?: string[];
  start_year?: number;
  end_year?: number;
  [key: string]: any;
}

interface PrismaMetrics {
  identification?: {
    records_identified: number;
  };
  screening?: {
    records_excluded_duplicates: number;
    records_after_duplicates_removed: number;
  };
  eligibility?: {
    full_text_assessed: number;
    full_text_excluded_semantic: number;
  };
  included?: {
    studies_included: number;
  };
}

export class LaTeXReviewGenerator {
  private genAI?: GoogleGenerativeAI;
  private model?: any;
  private bibtexGen: BibTeXGenerator;

  constructor() {
    this.bibtexGen = new BibTeXGenerator();

    // Configure Gemini
    if (config.GEMINI_API_KEY) {
      this.genAI = new GoogleGenerativeAI(config.GEMINI_API_KEY);
      this.model = this.genAI.getGenerativeModel({ model: 'gemini-pro' });
    } else {
      logger.warn('GEMINI_API_KEY not configured, AI generation disabled');
    }
  }

  private generateLatexHeader(title: string, author: string = 'Anonymous'): string {
    return `\\documentclass[12pt,a4paper]{article}

% Packages
\\usepackage[utf8]{inputenc}
\\usepackage[T1]{fontenc}
\\usepackage{geometry}
\\usepackage{cite}
\\usepackage{url}
\\usepackage{hyperref}
\\usepackage{graphicx}
\\usepackage{booktabs}
\\usepackage{longtable}

% Page geometry
\\geometry{
    a4paper,
    left=2.5cm,
    right=2.5cm,
    top=2.5cm,
    bottom=2.5cm
}

% Hyperref setup
\\hypersetup{
    colorlinks=true,
    linkcolor=blue,
    citecolor=blue,
    urlcolor=blue
}

% Title information
\\title{${title}}
\\author{${author}}
\\date{\\today}

\\begin{document}

\\maketitle

\\begin{abstract}
This systematic literature review was automatically generated using LitRevTool.
The review analyzes papers collected from Google Scholar based on specific search criteria.
The PRISMA methodology was followed to ensure systematic and transparent reporting.
\\end{abstract}

\\tableofcontents
\\newpage
`;
  }

  private generateIntroductionSection(searchCriteria: SearchCriteria): string {
    const keywordsInc = (searchCriteria.keywords_include || []).join(', ');
    const keywordsExc = (searchCriteria.keywords_exclude || []).join(', ');
    const startYear = searchCriteria.start_year || 'N/A';
    const endYear = searchCriteria.end_year || 'N/A';

    let intro = `\\section{Introduction}

This systematic literature review examines the research literature based on the following search strategy:

\\subsection{Search Criteria}

\\begin{itemize}
    \\item \\textbf{Inclusion Keywords}: ${keywordsInc}
`;

    if (keywordsExc) {
      intro += `    \\item \\textbf{Exclusion Keywords}: ${keywordsExc}\n`;
    }

    intro += `    \\item \\textbf{Year Range}: ${startYear} to ${endYear}
    \\item \\textbf{Database}: Google Scholar
\\end{itemize}

\\subsection{Methodology}

This review follows the PRISMA (Preferred Reporting Items for Systematic Reviews and Meta-Analyses) guidelines.
The search was conducted using LitRevTool, which overcomes Google Scholar's limitations by splitting searches by year.

`;
    return intro;
  }

  private generatePrismaSection(prismaMetrics?: PrismaMetrics): string {
    if (!prismaMetrics) return '';

    const identification = prismaMetrics.identification?.records_identified || 0;
    const duplicatesRemoved = prismaMetrics.screening?.records_excluded_duplicates || 0;
    const afterDedup = prismaMetrics.screening?.records_after_duplicates_removed || 0;
    const assessed = prismaMetrics.eligibility?.full_text_assessed || 0;
    const excluded = prismaMetrics.eligibility?.full_text_excluded_semantic || 0;
    const included = prismaMetrics.included?.studies_included || 0;

    return `\\section{PRISMA Flow}

The following table summarizes the systematic review process according to PRISMA 2020 guidelines:

\\begin{table}[h]
\\centering
\\begin{tabular}{ll}
\\toprule
\\textbf{Stage} & \\textbf{Count} \\\\
\\midrule
\\textbf{Identification} & \\\\
Records identified from database searching & ${identification} \\\\
\\midrule
\\textbf{Screening} & \\\\
Duplicate records removed & ${duplicatesRemoved} \\\\
Records after duplicates removed & ${afterDedup} \\\\
\\midrule
\\textbf{Eligibility} & \\\\
Full-text articles assessed & ${assessed} \\\\
Full-text articles excluded & ${excluded} \\\\
\\midrule
\\textbf{Included} & \\\\
\\textbf{Studies included in review} & \\textbf{${included}} \\\\
\\bottomrule
\\end{tabular}
\\caption{PRISMA 2020 flow summary}
\\label{tab:prisma}
\\end{table}

`;
  }

  private async generateResultsSection(papers: Paper[]): Promise<string> {
    let section = `\\section{Results}

A total of ${papers.length} papers were identified and included in this review.

\\subsection{Overview of Included Studies}

The following table provides an overview of the included studies:

\\begin{longtable}{p{0.4\\textwidth}p{0.3\\textwidth}p{0.15\\textwidth}p{0.1\\textwidth}}
\\toprule
\\textbf{Title} & \\textbf{Authors} & \\textbf{Source} & \\textbf{Year} \\\\
\\midrule
\\endfirsthead

\\multicolumn{4}{c}{\\textit{(Continued)}} \\\\
\\toprule
\\textbf{Title} & \\textbf{Authors} & \\textbf{Source} & \\textbf{Year} \\\\
\\midrule
\\endhead

\\bottomrule
\\endfoot

`;

    // Add papers to table (limit to prevent huge tables)
    const papersToShow = papers.slice(0, 100);
    for (const paper of papersToShow) {
      const title = this.escapeLatex(paper.title || '');
      const authors = this.escapeLatex((paper.authors || '').substring(0, 50));
      const source = this.escapeLatex((paper.source || '').substring(0, 30));
      const year = paper.year || 'N/A';

      section += `${title} & ${authors} & ${source} & ${year} \\\\\n`;
    }

    section += `\\end{longtable}\n\n`;

    // Add AI-generated analysis if available
    if (this.model && papers.length > 0) {
      try {
        logger.info('LaTeX: Generating AI analysis of papers...');
        const analysis = await this.generateAIAnalysis(papers);
        section += `\\subsection{Key Findings}\n\n${analysis}\n\n`;
      } catch (error) {
        logger.error('LaTeX: Failed to generate AI analysis', error);
      }
    }

    return section;
  }

  private async generateAIAnalysis(papers: Paper[]): Promise<string> {
    if (!this.model) return '';

    try {
      // Prepare paper summaries (limit to avoid token limits)
      const paperSummaries = papers.slice(0, 20).map((p, i) => {
        return `${i + 1}. ${p.title} (${p.year || 'N/A'})${p.abstract ? `: ${p.abstract.substring(0, 200)}...` : ''}`;
      });

      const prompt = `You are a research analyst. Based on the following academic papers, provide a brief analysis (3-4 paragraphs) of the key findings, trends, and themes. Format the output in LaTeX-compatible plain text (no special LaTeX commands needed, just paragraphs):

Papers:
${paperSummaries.join('\n\n')}

Provide a concise academic analysis of these papers:`;

      const result = await this.model.generateContent(prompt);
      const response = await result.response;
      return response.text();
    } catch (error) {
      logger.error('LaTeX: AI analysis error', error);
      return '';
    }
  }

  private generateConclusionSection(papers: Paper[]): string {
    return `\\section{Conclusion}

This systematic literature review identified ${papers.length} relevant studies based on the specified search criteria.
The review provides a comprehensive overview of the current state of research in this area.

Future research directions and gaps identified in the literature should be addressed to advance the field further.

`;
  }

  private generateReferencesSection(): string {
    return `\\section{References}

\\bibliographystyle{plain}
\\bibliography{references}

`;
  }

  private escapeLatex(text: string): string {
    if (!text) return '';

    return text
      .replace(/\\/g, '\\textbackslash{}')
      .replace(/[&%$#_{}]/g, (match) => '\\' + match)
      .replace(/~/g, '\\textasciitilde{}')
      .replace(/\^/g, '\\textasciicircum{}');
  }

  /**
   * Generate a complete LaTeX systematic literature review
   */
  async generateReview(options: {
    papers: Paper[];
    searchCriteria: SearchCriteria;
    prismaMetrics?: PrismaMetrics;
    title?: string;
    author?: string;
    outputPath?: string;
  }): Promise<string> {
    const { papers, searchCriteria, prismaMetrics, title, author, outputPath } = options;

    logger.info(`LaTeX: Generating review for ${papers.length} papers`);

    const reviewTitle = title || 'Systematic Literature Review';
    const reviewAuthor = author || 'LitRevTool';

    const sections: string[] = [];

    // Header
    sections.push(this.generateLatexHeader(reviewTitle, reviewAuthor));

    // Introduction
    sections.push(this.generateIntroductionSection(searchCriteria));

    // PRISMA
    if (prismaMetrics) {
      sections.push(this.generatePrismaSection(prismaMetrics));
    }

    // Results
    sections.push(await this.generateResultsSection(papers));

    // Conclusion
    sections.push(this.generateConclusionSection(papers));

    // References
    sections.push(this.generateReferencesSection());

    // Footer
    sections.push('\\end{document}');

    const fullContent = sections.join('\n');

    // Save to file if path provided
    if (outputPath) {
      try {
        fs.writeFileSync(outputPath, fullContent, 'utf-8');
        logger.info(`LaTeX file saved to ${outputPath}`);
      } catch (error) {
        logger.error('Error saving LaTeX file:', error);
      }
    }

    return fullContent;
  }
}

/**
 * Convenience function to generate LaTeX review
 */
export async function generateLatexReview(options: {
  papers: Paper[];
  searchCriteria: SearchCriteria;
  prismaMetrics?: PrismaMetrics;
  title?: string;
  author?: string;
  outputPath?: string;
}): Promise<string> {
  const generator = new LaTeXReviewGenerator();
  return generator.generateReview(options);
}
